htlcswitch/link.go

```diff
@@ -321,6 +321,19 @@ type ChannelLinkConfig struct {
 	// DeliveryAddr is the address to use in Shutdown if an upfront script
 	// has not been set.
 	DeliveryAddr lnwire.DeliveryAddress
+
+	// NotifySendingShutdown is used by the link to notify an outside
+	// subsystem that Shutdown is about to be sent. This is not used during
+	// retransmission. It takes a quit chan that is used as a precaution to
+	// avoid deadlock.
+	NotifySendingShutdown func(req *ChanClose, quit chan struct{})
```

Comment:
I think this flow should be reversed still: the link should gain an async way (which already exists) to notify the chan closer state machine that it can enter the shutdown flow.